### PR TITLE
Enable arm64 optimizations that exist for power/x86

### PR DESCRIPTION
**Pull request by AGSaidi**
Enable a set of optimizations that exist already for power and x86 for aarch64/arm64 systems.

Passes make check after these changes.

Running the string benchmarks the unaligned access change improves performance
by an average of 1.04x, min .96x, max 1.21x, median 1.01x

The gc optimization improves benchmark/gc/hash1 by 5%

The vm_exec changes make a massive difference on some benchmarks (e.g. 1.38x).

### CODE REVIEW

File: vm_exec.c [hunk 1]
```diff
@@ -93,6 +96,11 @@ vm_exec_core(rb_execution_context_t *ec, VALUE initial)
     DECL_SC_REG(rb_control_frame_t *, cfp, "15");
 #define USE_MACHINE_REGS 1
 
+#elif defined(__GNUC__) && defined(__aarch64__)
+    DECL_SC_REG(const VALUE *, pc, "19");
+    DECL_SC_REG(rb_control_frame_t *, cfp, "20");
+#define USE_MACHINE_REGS 1
+
```
Comment (shyouhei) [inline vm_exec.c:103]: Does this really benefit?  We know that recent compilers are smarter than we wrote those sibling codes.  Read more: https://bugs.ruby-lang.org/issues/12225

cc @nurse

Comment (AGSaidi) [reply vm_exec.c:103]: @shyouhei the only changes between compare-ruby and built-ruby in the number in the commit message above are the two hunks in vm_exec.c. I'm happy to run other benchmarks if you'd like, but it appears to improve substantially. Double checked my result again by removing all diffs and comparing to the ruby I built prior to my patches. The results were +-2% and then reapplied these two hunks and re-ran again, and observed the improvements here (up to 1.38x).

Comment (nurse) [reply vm_exec.c:103]: As far as I remember, there're another example with clang which says it's still effective.
And [the commit](https://github.com/ruby/ruby/pull/3393/commits/c985b8c6868a380e44e285368af4a4f414ce3309) comment says 1.2x seems worth introducing this change.

Comment (shyouhei) [reply vm_exec.c:103]: OK then, we need to investigate what is going on but this pull request can be a separate thing.

Comment (AGSaidi) [reply vm_exec.c:103]: @nurse  anything else you'd like to see before you merge?

Comment (nurse) [reply vm_exec.c:103]: I think this is OK to merge.
@shyouhei Do you have another topic?

Comment (shyouhei) [reply vm_exec.c:103]: @nurse No, it is LTGM.

**Pull request merged**